systemvm/patches/debian/config/opt/cloud/bin/passwd_server_ip.py

```diff
@@ -0,0 +1,204 @@
+    try:
+        f = open(getPasswordFile(), "w")
+        for ip in passMap.keys():
+            f.write("%s=%s\n" % (ip, passMap[ip]))
+        f.close()
+    except IOError, e:
+        syslog.syslog("serve_password: Unable to save to password file %s" % e)
+    finally:
+        lock.release()
+
+def getPassword(ip):
+    global passMap
+    if ip in passMap:
+        return passMap[ip]
+    return None
+
+def setPassword(ip, password):
+    global passMap, lock
+    if ip == None or ip == "" or password == None or password == "":
+        return
+    lock.acquire()
+    try:
+        passMap[ip] = password
+    finally:
+        lock.release()
+
+def removePassword(ip):
+    global passMap, lock
+    lock.acquire()
+    try:
+        if ip in passMap:
+            del passMap[ip]
+    finally:
+        lock.release()
+
+
+class ThreadedHTTPServer(ThreadingMixIn, HTTPServer):
+    pass
+
+
+class PasswordRequestHandler(BaseHTTPRequestHandler):
+    server_version = "CloudStack Password Server"
+    sys_version = "4.x"
+    def do_GET(self):
+        self.send_response(200)
+        self.send_header("Content-type", "text/plain")
+        self.send_header("Server", "CloudStack Password Server")
+        self.end_headers()
+        requestType = self.headers.get('DomU_Request')
+        clientAddress = self.client_address[0]
+        if requestType == "send_my_password":
+            password = getPassword(clientAddress)
+            if not password:
+                syslog.syslog("serve_password: requested password not found for %s" % clientAddress)
+            else:
+                self.wfile.write(password)
+                syslog.syslog("serve_password: password sent to %s" % clientAddress)
+        elif requestType == "saved_password":
+            removePassword(clientAddress)
+            savePasswordFile()
+            self.wfile.write("saved_password")
+            syslog.syslog("serve_password: saved_password ack received from %s" % clientAddress)
+        else:
+            self.send_response(400)
+            self.wfile.write("bad_request")
+            syslog.syslog("serve_password: bad_request from IP %s" % clientAddress)
+        return
+
+    def do_POST(self):
+        form = cgi.FieldStorage(
+                    fp=self.rfile,
+                    headers=self.headers,
+                    environ={'REQUEST_METHOD':'POST',
+                             'CONTENT_TYPE':self.headers['Content-Type'],
+                    })
+        self.send_response(200)
+        self.end_headers()
+        clientAddress = self.client_address[0]
+        if not (clientAddress == "localhost" or clientAddress == "127.0.0.1"):
+            syslog.syslog("serve_password: non-localhost IP trying to save password: %s" % clientAddress)
+            self.send_response(403)
+            return
+        if "ip" not in form.keys() or "password" not in form.keys() or "token" not in form.keys() or self.headers.get('DomU_Request') != "save_password":
```

Comment:
With dictionaries you can directly check for membership with `if "ip" not in form`.